porch/controllers/packagevariantsets/pkg/controllers/packagevariantset/packagevariantset_controller.go

```diff
@@ -90,349 +90,161 @@ func (r *PackageVariantSetReconciler) Reconcile(ctx context.Context, req ctrl.Re
-		if err != nil {
-			return nil, fmt.Errorf("error converting repo to RNode: %v", err)
-		}
-		downstreamPackageName, err := r.getDownstreamPackageName(target.PackageName, upstreamPackageName, repoAsRNode)
-		if err != nil {
-			return nil, err
+				for _, pn := range pns {
+					result = append(result, pvContext{
+						template:       target.Template,
+						repoDefault:    rt.Name,
+						packageDefault: pn,
+					})
+				}
+			}
+			continue
 		}
-		result = append(result, &pkgvarapi.Downstream{
-			Repo:    repo.Name,
-			Package: downstreamPackageName,
-		})
 
-	}
-	return result, nil
-}
+		objSel := target.ObjectSelector
+		if target.RepositorySelector != nil {
+			// a label selector against a set of repositories
+			// equivlanet to object selector with apiVersion/kind pre-set
 
-func (r *PackageVariantSetReconciler) objectSet(target *api.Target,
-	upstreamPackageName string,
-	selectedObjects map[resid.ResId]*yaml.RNode) ([]*pkgvarapi.Downstream, error) {
-	var result []*pkgvarapi.Downstream
-	for _, obj := range selectedObjects {
-		downstreamPackageName, err := r.getDownstreamPackageName(target.PackageName,
-			upstreamPackageName, obj)
-		if err != nil {
-			return nil, err
-		}
-		repo, err := r.fetchValue(target.Objects.RepoName, obj)
-		if err != nil {
-			return nil, err
-		}
-		if repo == "" {
-			return nil, fmt.Errorf("error evaluating repo name: received empty string")
+			objSel = &api.ObjectSelector{
+				LabelSelector: *target.RepositorySelector,
+				APIVersion:    configapi.TypeRepository.APIVersion(),
+				Kind:          configapi.TypeRepository.Kind,
+			}
 		}
-		result = append(result, &pkgvarapi.Downstream{
-			Package: downstreamPackageName,
-			Repo:    repo,
-		})
-	}
-	return result, nil
-}
-
-func (r *PackageVariantSetReconciler) getSelectedObjects(ctx context.Context, selectors []api.Selector) (map[resid.ResId]*yaml.RNode, error) {
-	selectedObjects := make(map[resid.ResId]*yaml.RNode) // this is a map to prevent duplicates
-
-	for _, selector := range selectors {
+		// a selector against a set of arbitrary objects
 		uList := &unstructured.UnstructuredList{}
-		group, version := resid.ParseGroupVersion(selector.APIVersion)
+		group, version := resid.ParseGroupVersion(objSel.APIVersion)
 		uList.SetGroupVersionKind(schema.GroupVersionKind{
 			Group:   group,
 			Version: version,
-			Kind:    selector.Kind,
+			Kind:    objSel.Kind,
 		})
 
-		opts := []client.ListOption{client.InNamespace(selector.Namespace)}
-		if selector.Labels != nil {
-			labelSelector, err := metav1.LabelSelectorAsSelector(selector.Labels)
-			if err != nil {
-				return nil, err
-			}
-			opts = append(opts, client.MatchingLabelsSelector{Selector: labelSelector})
+		opts := []client.ListOption{client.InNamespace(pvs.Namespace)}
+		labelSelector, err := metav1.LabelSelectorAsSelector(&objSel.LabelSelector)
+		if err != nil {
+			return nil, err
 		}
+		opts = append(opts, client.MatchingLabelsSelector{Selector: labelSelector})
```

Comment:
if `objSel.LabelSelector` is empty, it should presumably match everything. Is that what happens?

Reply:
Good question. I tried it out, and yes, that's what happens. But while trying it out I found another bug, so pushed a fix for that!